ui/src/shared/actions/predicates.ts

```diff
@@ -153,6 +182,47 @@ export const deleteWithPredicate = params => async (
   }
 }
 
+export const executePreviewQuery = (query: string) => async (
+  dispatch,
+  getState: GetState
+) => {
+  dispatch(setPreviewStatus(RemoteDataState.Loading))
+  try {
+    const orgID = getState().orgs.org.id
+
+    const variableAssignments = getVariableAssignments(getState())
+    const windowVars = getWindowVars(query, variableAssignments)
+    const extern = buildVarsOption([...variableAssignments, ...windowVars])
+    const result = await runQuery(orgID, query, extern).promise
+
+    if (result.type === 'UNKNOWN_ERROR') {
+      throw new Error(result.message)
+    }
+
+    if (result.type === 'RATE_LIMIT_ERROR') {
+      dispatch(notify(rateLimitReached(result.retryAfter)))
+
+      throw new Error(result.message)
+    }
+
+    if (result.didTruncate) {
+      dispatch(notify(resultTooLarge(result.bytesRead)))
+    }
+
+    checkQueryResult(result.csv)
+
+    const files = [result.csv]
+    dispatch(setFiles(files))
+  } catch (e) {
+    if (e.name === 'CancellationError') {
```

Comment:
should still set the preview status to something here

Reply:
Agreed. We should discuss with design about what feedback beyond a notification we can return to the user